### PR TITLE
James 1838 Clarify shared mailbox access

### PR DESCRIPTION
**Pull request by chibenwa**
I reworked MailboxPath generation and improved it a bit. Someone can not access any mailbox he do not own.

About paths, if my username is **chibenwa** : 

 - Path *toto* will point to *#private:chibenwa:toto*
 - Path *#private.toto* will also point to *#private:chibenwa:toto* 

With #private username is implied to be yours (previous behaviour)

 - Path *#other.toto* will point to *#other:null:toto*

With shared mailboxes, username component of MailboxPath is implied to be null.

Question : Do we modify MailboxPath to have Optional user ? It is some serious refactoring, so I want to have your opinion before doing it. We may delay it to shared mailbox implementation.

I keep other questions for shared mailbox implementation ;-)

### CODE REVIEW

**Comment by aduprat** [comment]
What happens if there is 'null' user?
I think we may introduce a keyword for shared mailboxes; then it will make sense to have user optional.

**Comment by chibenwa** [comment]
Value is null @aduprat not "null", then no troubles.

File: protocols/imap/src/main/java/org/apache/james/imap/main/DeniedSharedMailboxException.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.imap.main;
+
+public class DeniedSharedMailboxException extends RuntimeException {
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/main/DeniedSharedMailboxException.java]: Deny what ? Access ?

File: protocols/imap/src/test/java/org/apache/james/imap/main/PathConverterTest.java [hunk 2]
```diff
@@ -71,6 +75,7 @@ public void buildPathShouldAcceptRelativeMailboxName() {
             .isEqualTo(new MailboxPath(MailboxConstants.USER_NAMESPACE, USERNAME, mailboxName));
     }
 
+    @Ignore
```
Comment (mbaechler) [inline protocols/imap/src/test/java/org/apache/james/imap/main/PathConverterTest.java]: @Ignore should come with a message to explain why it's ignored

File: protocols/imap/src/main/java/org/apache/james/imap/api/display/HumanReadableText.java [hunk 3]
```diff
@@ -158,6 +158,8 @@ public static final HumanReadableText permanentFlags(Flags flags) {
     public static final HumanReadableText QRESYNC_CLOSED = new HumanReadableText("org.apache.james.imap.QRESYNC_CLOSED", "");
     public static final HumanReadableText QRESYNC_VANISHED_WITHOUT_CHANGEDSINCE = new HumanReadableText("org.apache.james.imap.QRESYNC_VANISHED_WITHOUT_CHANGEDSINCE", "VANISHED used without CHANGEDSINCE");
 
+    public static final HumanReadableText DENIED_SHARED_MAILBOX = new HumanReadableText("org.apache.james.imap.DENIED_SHARED_MAILBOX", "You can not access a mailbox that do not belong to you");
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/api/display/HumanReadableText.java]: s/do/does

File: protocols/imap/src/main/java/org/apache/james/imap/main/PathConverter.java [hunk 4]
```diff
@@ -38,41 +45,48 @@ public PathConverter(ImapSession session) {
...
-            namespace = MailboxConstants.USER_NAMESPACE;
-            name = mailboxName;
+            return buildRelativePath(mailboxName);
         }
-        String user = null;
-        // we only use the user as part of the MailboxPath if its a private user
-        // namespace
+    }
+
+    private boolean isAbsolute(String mailboxName) {
```
Comment (chibenwa) [inline protocols/imap/src/main/java/org/apache/james/imap/main/PathConverter.java]: Absolute path have a namespace specified.

Relative path concerns user mailboxes.

Hence the check to namespace char.

**Pull request merged**